lib/wasi/src/syscalls/wasi/fd_read.rs

```diff
@@ -0,0 +1,364 @@
+    wasi_try_ok!(ctx.data().clone().process_signals_and_exit(&mut ctx)?);
+
+    let mut env = ctx.data();
+    let state = env.state.clone();
+    let inodes = state.inodes.clone();
+
+    let fd_entry = wasi_try_ok!(state.fs.get_fd(fd));
+    let is_stdio = fd_entry.is_stdio;
+
+    let bytes_read = {
+        if is_stdio == false {
+            if !fd_entry.rights.contains(Rights::FD_READ) {
+                // TODO: figure out the error to return when lacking rights
+                return Ok(Errno::Access);
+            }
+        }
+
+        let is_non_blocking = fd_entry.flags.contains(Fdflags::NONBLOCK);
+        let inode_idx = fd_entry.inode;
+
+        let max_size = {
+            let memory = env.memory_view(&ctx);
+            let iovs_arr = wasi_try_mem_ok!(iovs.slice(&memory, iovs_len));
+            let mut max_size = 0usize;
+            for iovs in iovs_arr.iter() {
+                let iovs = wasi_try_mem_ok!(iovs.read());
+                let buf_len: usize =
+                    wasi_try_ok!(iovs.buf_len.try_into().map_err(|_| Errno::Overflow));
+                max_size += buf_len;
+            }
+            max_size
+        };
+
+        let (bytes_read, can_update_cursor) = {
+            let inodes = inodes.read().unwrap();
+            let inode = &inodes.arena[inode_idx];
+            let mut guard = inode.write();
+            match guard.deref_mut() {
+                Kind::File { handle, .. } => {
+                    if let Some(handle) = handle {
+                        let handle = handle.clone();
+                        drop(inode);
+                        drop(guard);
+                        drop(inodes);
+
+                        let data = wasi_try_ok!(__asyncify(
+                            &mut ctx,
+                            if is_non_blocking {
+                                Some(Duration::ZERO)
+                            } else {
+                                None
+                            },
+                            async move {
+                                let mut handle = handle.write().unwrap();
+                                if is_stdio == false {
+                                    handle
+                                        .seek(std::io::SeekFrom::Start(offset as u64))
+                                        .await
+                                        .map_err(map_io_err)?;
+                                }
+
+                                let mut data = Vec::with_capacity(max_size);
+                                unsafe { data.set_len(max_size) };
+                                let amt = handle.read(&mut data[..]).await.map_err(|err| {
+                                    let err = From::<std::io::Error>::from(err);
+                                    match err {
+                                        Errno::Again => {
+                                            if is_stdio {
+                                                Errno::Badf
```

Comment:
@ptitSeb helped fix a test that was failing by adding this check.

I assume this will break consumers, because reading from an empty pipe is not a wrong thing to do and shouldn't return an error. 

I'll investigate if this is problematic for dash/bash etc.

Reply:
@john-sharratt

Reply:
Check the rust sources: https://github.com/rust-lang/rust/blob/415d8fcc3e17f8c1324a81cf2aa7127b4fcfa32e/library/std/src/io/stdio.rs#L96-L117

You'll see that the ebadf error is the expected and return an Ok (with the `handle_ebadf` function).

Reply:
:+1: